### PR TITLE
Support multiple supervision edge types in ABLP Loader

### PR DESCRIPTION
**Pull request by kmontemayor2-sc**
**Scope of work done**

Added support for multiple supervision edge types.

There are two primary user-facing changes here:

1.  [Enable](https://github.com/Snapchat/GiGL/blob/ffe3d601836eb6b6f2af0c097628849d85635d92/python/gigl/distributed/dist_ablp_neighborloader.py#L61) multiple supervision edges types to be passed into the loader 
2. When more than one supervision edge types are provided, make y_positive/y_negative a `dict[EdgeType, dict[int, Tensor]]` [ex](https://github.com/Snapchat/GiGL/blob/ffe3d601836eb6b6f2af0c097628849d85635d92/python/tests/unit/distributed/dist_ablp_neighborloader_test.py#L790-L798). 

And then a lot of piping to go from 1 -> 2...

Where is the documentation for this feature?: 

In [docstring](https://github.com/Snapchat/GiGL/blob/ffe3d601836eb6b6f2af0c097628849d85635d92/python/gigl/distributed/dist_ablp_neighborloader.py#L117-L121) for ABLP loader

Did you add automated tests or write a test plan?

Added unit tests

***Updated Changelog.md?*** Yes

***Ready for code review?:*** Yes

### CODE REVIEW

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 1]
```diff
@@ -170,6 +184,14 @@ def __init__(
         master_ip_address: str
         should_cleanup_distributed_context: bool = False
 
+        if supervision_edge_type is None:
+            self._supervision_edge_types: list[EdgeType] = []
+        elif isinstance(supervision_edge_type, list):
+            self._supervision_edge_types = supervision_edge_type
+        else:
+            self._supervision_edge_types = [supervision_edge_type]
+        del supervision_edge_type
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/dist_ablp_neighborloader.py:193]: No reason, just trying to prevent mis-using this later (instead of self._supervision_edge_types which we are using)

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 2]
```diff
@@ -117,13 +127,17 @@ def __init__(
...
-            local_process_rank (int): The local rank of the current process within a node.
-            local_process_world_size (int): The total number of processes within a node.
             input_nodes (Optional[torch.Tensor, tuple[NodeType, torch.Tensor]]):
                 Indices of seed nodes to start sampling from.
                 If set to `None` for homogeneous settings, all nodes will be considered.
                 In heterogeneous graphs, this flag must be passed in as a tuple that holds
                 the node type and node indices. (default: `None`)
+            supervision_edge_type (Optional[Union[EdgeType, list[EdgeType]]]):
+                The edge type(s) to use for supervision.
+                Must be None iff the dataset is labeled homogeneous.
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/dist_ablp_neighborloader.py:137]: Good call! throwing now if an empty list is provided.

[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/dist_ablp_neighborloader.py:137]: also added some tests for empty inputs :P

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 3]
```diff
@@ -497,82 +540,109 @@ def _get_labels(
             msg (SampleMessage): All possible results from a sampler, including subgraph data, features, and used defined metadata
         Returns:
             SampleMessage: Updated sample messsage with the label fields removed
-            torch.Tensor: Positive label ID tensor, where the ith row corresponds to the ith anchor node ID
-            Optional[torch.Tensor]: Negative label ID tensor, where the ith row corresponds to the ith anchor node ID, can be None if dataset has no negative labels
+            dict[EdgeType, torch.Tensor]: Positive label ID tensor, where the ith row corresponds to the ith anchor node ID
+            Optional[dict[EdgeType, torch.Tensor]]: Negative label ID tensor, where the ith row corresponds to the ith anchor node ID.
```
Comment (mkolodner-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:544]: Can we update these comments to reflect the heterogeneous nature here? Ditto elsewhere

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 4]
```diff
@@ -497,82 +540,109 @@ def _get_labels(
...
         """
-        metadata = {}
+        metadata: dict[str, torch.Tensor] = {}
+        positive_labels: dict[EdgeType, torch.Tensor] = {}
+        negative_labels: dict[EdgeType, torch.Tensor] = {}
         for k in list(msg.keys()):
-            if k.startswith("#META."):
-                meta_key = str(k[6:])
+            if k.startswith(f"#META.{POSITIVE_LABEL_METADATA_KEY}."):
```
Comment (mkolodner-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:550]: Adding some comments here would be helpful to better understand the logic here

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 5]
```diff
@@ -497,82 +540,109 @@ def _get_labels(
...
+
+                # Gets the indexes of the items in local_node_to_global_node which match any of the positive labels for the current anchor node
+                output_positive_labels[
+                    label_edge_type_to_message_passing_edge_type(edge_type)
+                ][local_anchor_node_id] = torch.nonzero(positive_mask)[:, 0].to(
+                    self.to_device
+                )
+                # Shape [X], where X is the number of indexes in the original local_node_to_global_node which match a node in the positive labels for the current anchor node
+
         if negative_labels is not None:
```
Comment (mkolodner-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:620]: Negative labels cannot be none anymore?

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 6]
```diff
@@ -497,82 +540,109 @@ def _get_labels(
...
+                    )  # shape [N, M], where N is the number of nodes and M is the number of negative labels for the current anchor node
+
+                    # Gets the indexes of the items in local_node_to_global_node which match any of the negative labels for the current anchor node
+                    output_negative_labels[
+                        label_edge_type_to_message_passing_edge_type(edge_type)
+                    ][local_anchor_node_id] = torch.nonzero(negative_mask)[:, 0].to(
+                        self.to_device
+                    )
+                    # Shape [X], where X is the number of indexes in the original local_node_to_global_node which match a node in the negative labels for the current anchor node
+        if len(output_positive_labels) == 1:
```
Comment (mkolodner-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:637]: Could positive labels ever be len(0) here? Should we raise a ValueError if this happens?

File: python/gigl/distributed/dist_neighbor_sampler.py [hunk 7]
```diff
@@ -35,48 +39,42 @@ async def _sample_from_nodes(
...
+        metadata: dict[str, torch.Tensor] = {}
+        input_seeds_builder: dict[
+            Union[str, NodeType], list[torch.Tensor]
+        ] = defaultdict(list)
+        input_seeds_builder[input_type].append(input_seeds)
+        for edge_type, label_tensor in inputs.positive_label_by_edge_types.items():
+            input_seeds_builder[edge_type[label_edge_index]].append(
+                label_tensor[label_tensor != PADDING_NODE].to(self.device)
+            )
+            metadata[
```
Comment (mkolodner-sc) [inline python/gigl/distributed/dist_neighbor_sampler.py:59]: Rather than having to do all of this string parsing here so that each edge type is a different field in the metadata, could we just store the edge type information in the value itself (i.e. store the value as a dict[edge_type, label_tensor])?

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 8]
```diff
@@ -291,22 +319,41 @@ def __init__(
...
-        positive_labels, negative_labels = get_labels_for_anchor_nodes(
-            dataset=dataset,
-            node_ids=curr_process_nodes,
-            positive_label_edge_type=self._positive_label_edge_type,
-            negative_label_edge_type=self._negative_label_edge_type,
+
+        self._positive_label_edge_types: list[EdgeType] = []
+        self._negative_label_edge_types: list[EdgeType] = []
+        all_positive_labels: dict[EdgeType, torch.Tensor] = {}
+        all_negative_labels: dict[EdgeType, torch.Tensor] = {}
```
Comment (mkolodner-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:332]: Can you update these names to be more reflective of their dictionary typing?

File: python/gigl/distributed/sampler.py [hunk 9]
```diff
@@ -16,24 +19,28 @@ def __init__(
...
+        self._positive_label_by_edge_types = positive_label_by_edge_types
+        self._negative_label_by_edge_types = negative_label_by_edge_types
+
+    @property
+    def positive_label_by_edge_types(self) -> dict[EdgeType, torch.Tensor]:
+        return self._positive_label_by_edge_types
+
+    @property
+    def negative_label_by_edge_types(self) -> dict[EdgeType, torch.Tensor]:
+        return self._negative_label_by_edge_types
```
Comment (mkolodner-sc) [inline python/gigl/distributed/sampler.py:43]: qq: where are these used?

File: python/gigl/distributed/sampler.py [hunk 10]
```diff
@@ -45,9 +52,15 @@ def __getitem__(self, index: Union[torch.Tensor, Any]) -> "ABLPNodeSamplerInput"
         return ABLPNodeSamplerInput(
             node=self.node[index],
             input_type=self.input_type,
-            positive_labels=self.positive_labels[index],
-            negative_labels=self.negative_labels[index]
-            if self.negative_labels is not None
-            else None,
-            supervision_node_type=self.supervision_node_type,
+            positive_label_by_edge_types={
+                edge_type: self._positive_label_by_edge_types[edge_type][index]
```
Comment (mkolodner-sc) [inline python/gigl/distributed/sampler.py:56]: What is the behavior here if a given labeled edge type does not have that index?

**Review by mkolodner-sc — COMMENTED** [review]
Thanks Kyle! Did a pass here -- will look through testing files shortly

File: python/tests/unit/distributed/dist_ablp_neighborloader_test.py [hunk 11]
```diff
@@ -296,6 +313,113 @@ def _run_toy_heterogeneous_ablp(
...
+            dim=0,
+        )
+    assert hasattr(datum, "y_positive")
+    assert set(datum.y_positive.keys()) == set(
+        expected_positive_labels.keys()
+    ), f"{datum.y_positive.keys()} != {expected_positive_labels.keys()}"
+    for edge_type in datum.y_positive.keys():
+        _assert_labels(
+            datum[edge_type[0]].node,
+            datum[edge_type[2]].node,
```
Comment (yliu2-sc) [inline python/tests/unit/distributed/dist_ablp_neighborloader_test.py:361]: maybe assign 0/2 to in/out edge index variables so it's not magic

File: python/tests/unit/distributed/dist_ablp_neighborloader_test.py [hunk 12]
```diff
@@ -574,6 +698,260 @@ def test_toy_heterogeneous_ablp(
...
+                    graph_partition={},
+                    node_partition_book={},
+                ),
+                num_neighbors=[2, 2],
+                input_nodes=(NodeType("a"), torch.tensor([10])),
+                supervision_edge_type=[_A_TO_B, _A_TO_C],
+            ),
+        ]
+    )
+    def test_ablp_dataloder_invalid_inputs(
```
Comment (yliu2-sc) [inline python/tests/unit/distributed/dist_ablp_neighborloader_test.py:942]:
```suggestion
    def test_ablp_dataloader_invalid_inputs(
```

File: python/tests/unit/distributed/dist_ablp_neighborloader_test.py [hunk 13]
```diff
@@ -574,6 +698,260 @@ def test_toy_heterogeneous_ablp(
...
+                    edge_dir="out",
+                    graph_partition={},
+                    node_partition_book={},
+                ),
+                num_neighbors=[2, 2],
+                input_nodes=(NodeType("a"), torch.tensor([10])),
+                supervision_edge_type=None,
+            ),
+            param(
+                "Mutiple supervision edge types, homogeneous sampling",
```
Comment (yliu2-sc) [inline python/tests/unit/distributed/dist_ablp_neighborloader_test.py:926]: heterogeneous sampling?

**Review by yliu2-sc — APPROVED** [review]
In general LGTM

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 14]
```diff
@@ -170,6 +184,18 @@ def __init__(
         master_ip_address: str
         should_cleanup_distributed_context: bool = False
 
+        if supervision_edge_type is None:
+            self._supervision_edge_types: list[EdgeType] = []
+        elif isinstance(supervision_edge_type, list):
+            if not supervision_edge_type:
+                raise ValueError(
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/dist_ablp_neighborloader.py:193]: changed this up a bit, wdyt now?

File: python/gigl/distributed/dist_neighbor_sampler.py [hunk 15]
```diff
@@ -35,48 +40,56 @@ async def _sample_from_nodes(
...
+            input_seeds_builder[edge_type[label_edge_index]].append(
+                filtered_label_tensor
+            )
+            # Update the metadata per negative label edge type.
+            # We do this because GLT only supports dict[str, torch.Tensor] for metadata.
+            metadata[
+                f"{NEGATIVE_LABEL_METADATA_KEY}.{str(tuple(edge_type))}"
+            ] = label_tensor
+        input_nodes: dict[Union[str, NodeType], torch.Tensor] = {
+            node_type: torch.cat(seeds, dim=0).to(self.device)
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/dist_neighbor_sampler.py:81]: Hmmm, think doing so is a perf optimization right? WDYT if I leave a TODO?

The way I see it is: we should only do this if we have multiple supervision edge types with *different* relations but the same src/dst nodes (or for "homogeneous" graphs), otherwise, we're doing the sort (which is slow) for no reason?

File: python/tests/unit/distributed/dist_ablp_neighborloader_test.py [hunk 16]
```diff
@@ -75,7 +89,8 @@ def _assert_labels(
     _assert_labels(node, y, expected)
 
     Args:
-        node (torch.Tensor): The node tensor, N, where N is the number of nodes in the batch,
+        src_node (torch.Tensor): The source (anchor) node tensor, [N], where N is the number of source nodes in the batch,
```
Comment (mkolodner-sc) [inline python/tests/unit/distributed/dist_ablp_neighborloader_test.py:92]: These names are misleading -- this is not the anchor nodes tensor for the batch, but the nodes in the batch with the same type as the anchor node type. Could we rename and update the description here? Ditto for the supervision nodes.

nit: we should probably not use `N` variable for both of these parameters here, since their values may be different.

Comment (mkolodner-sc) [reply python/tests/unit/distributed/dist_ablp_neighborloader_test.py:92]: Also, could we avoid using "src" and "dst" terms here? 

The current naming here is enforcing us to stick with the assumption that supervision edges are always directed outward, which we might want to change in the future.

File: python/tests/unit/distributed/dist_ablp_neighborloader_test.py [hunk 17]
```diff
@@ -132,9 +147,11 @@ def _run_distributed_ablp_neighbor_loader(
         expected_node,
         dim=0,
     )
-    _assert_labels(datum.node, datum.y_positive, expected_positive_labels)
+    _assert_labels(datum.node, datum.node, datum.y_positive, expected_positive_labels)
     if expected_negative_labels is not None:
-        _assert_labels(datum.node, datum.y_negative, expected_negative_labels)
+        _assert_labels(
+            datum.node, datum.node, datum.y_negative, expected_negative_labels
```
Comment (mkolodner-sc) [inline python/tests/unit/distributed/dist_ablp_neighborloader_test.py:153]: a comment here on why we pass in `datum.node` twice could be helpful

File: python/tests/unit/distributed/dist_ablp_neighborloader_test.py [hunk 18]
```diff
@@ -296,6 +313,115 @@ def _run_toy_heterogeneous_ablp(
...
+            datum[node_type].node,
+            expected_node[node_type],
+            dim=0,
+        )
+    assert hasattr(datum, "y_positive")
+    assert set(datum.y_positive.keys()) == set(
+        expected_positive_labels.keys()
+    ), f"{datum.y_positive.keys()} != {expected_positive_labels.keys()}"
+    src_index = 0
+    dst_index = 2
```
Comment (mkolodner-sc) [inline python/tests/unit/distributed/dist_ablp_neighborloader_test.py:359]: Same comment about avoiding use of "src" and "dst" terms here in favor of "anchor" and "supervision".

File: python/tests/unit/distributed/dist_ablp_neighborloader_test.py [hunk 19]
```diff
@@ -574,6 +700,260 @@ def test_toy_heterogeneous_ablp(
...
+                )
+                for etype, idx in edge_index.items()
+            },
+            partitioned_edge_features=None,
+            partitioned_node_features=None,
+            partitioned_negative_labels=None,
+            partitioned_positive_labels=None,
+            partitioned_node_labels=None,
+        )
+        dataset = DistDataset(rank=0, world_size=1, edge_dir="out")
```
Comment (mkolodner-sc) [inline python/tests/unit/distributed/dist_ablp_neighborloader_test.py:871]: Should we test "in" as well here?

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 20]
```diff
@@ -497,82 +544,121 @@ def _get_labels(
...
-        )
         return (msg, positive_labels, negative_labels)
 
     def _set_labels(
         self,
         data: Union[Data, HeteroData],
-        positive_labels: torch.Tensor,
-        negative_labels: Optional[torch.Tensor],
+        positive_labels: dict[EdgeType, torch.Tensor],
+        negative_labels: dict[EdgeType, torch.Tensor],
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/dist_ablp_neighborloader.py]: done.

File: python/tests/unit/distributed/dist_ablp_neighborloader_test.py (hunk 19 above)
Comment (kmontemayor2-sc) [reply python/tests/unit/distributed/dist_ablp_neighborloader_test.py]: done. found some bugs, thanks for asking for this!

File: python/tests/unit/distributed/dist_ablp_neighborloader_test.py [hunk 21]
```diff
@@ -75,9 +91,8 @@ def _assert_labels(
     _assert_labels(node, y, expected)
 
     Args:
-        node (torch.Tensor): The node tensor, N, where N is the number of nodes in the batch,
-        where the local IDs are the index of a node in the tensor,
-        and the global IDs are the values of the tensor.
+        anchor_nodes (torch.Tensor): The source (anchor) node tensor, [N], where N is the number of supervision nodes in the batch,
+        supervision_nodes (torch.Tensor): The destination (supervision) node tensor, [M], where M is the number of supervision nodes in the batch,
```
Comment (mkolodner-sc) [inline python/tests/unit/distributed/dist_ablp_neighborloader_test.py:95]: suggested rewording:
```
anchor_nodes (torch.Tensor): Tensor of nodes in the graph with the same type as the anchor node, shape [N] where N is the number of nodes in the batch with this node type
supervision_nodes (torch.Tensor): Tensor of nodes in the graph with the same type as the supervision node, shape [M] where M is the number of nodes in the batch with this node type

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 22]
```diff
@@ -497,82 +545,127 @@ def _get_labels(
...
+            dict
+        )
+        output_negative_labels: dict[EdgeType, dict[int, torch.Tensor]] = defaultdict(
+            dict
+        )
+        # Since GLT swaps src/dst for edge_dir = "out",
+        # and GiGL assumes that supervision edge types are always (anchor_node_type, to, supervision_node_type),
+        # we need to index into supervision edge types accordingly.
+        # edge_index = 0 if self.edge_dir == "in" else 2
+        edge_index = 2
```
Comment (mkolodner-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:624]: Should we remove outdated comment and commented code here? Also add explanation for why this is always 2 for now would be helpful

**Review by mkolodner-sc — APPROVED** [review]
Thanks a lot Kyle! LGTM provided tests pass

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 23]
```diff
@@ -497,82 +546,139 @@ def _get_labels(
...
+                DEFAULT_HOMOGENEOUS_NODE_TYPE
+            ] = data.node
+        output_positive_labels: dict[EdgeType, dict[int, torch.Tensor]] = defaultdict(
+            dict
+        )
+        output_negative_labels: dict[EdgeType, dict[int, torch.Tensor]] = defaultdict(
+            dict
+        )
+        # We always have supervision edge types of the form (anchor_node_type, to, supervision_node_type)
+        # So we can index into the edge type accordingly.
```
[the remark this answers is not in the document]
Comment (mkolodner-sc) [reply python/gigl/distributed/dist_ablp_neighborloader.py:636]: Agreed -- I was thinking that we could remove the edge flipping altogether. I had a POC for the work needed here, but we should probably first align at some point what the user should be providing across the the different strategies (SSL, UDL) and edge directions in this new paradigm.
